lib/common_test.sh

```diff
@@ -28,28 +30,67 @@ readonly TESTDATA="$( cd "$(dirname "$0")" && pwd )/testdata"
 #       - add `set -o nounset`
 #
 #       We can do that when all the things we source do not rely on unset
-#       varaibales and ignoring errors. This will require quite some
+#       variables and ignoring errors. This will require quite some
 #       refactoring, so this is the best we can do for now.
 set -o errexit
 set -o pipefail
 
 TEST_run_stateful() {
-  tmpDir="$( mktemp -d )"
-  trap 'rm -rf -- "$tmpDir"' EXIT
-
-  # override some vars and func to not clutter output
-  common::timestamp() { :; }
-  # shellcheck disable=SC2034
-  PROGSTATE="${tmpDir}/whats-a-progstate-even.txt" \
-    LOGFILE="${tmpDir}/some-log-file.log" \
-    HR='' \
-    TPUT[BOLD]='' \
-    TPUT[OFF]=''
+  test_scaffold
 
   assert_equal_content \
     <( common::run_stateful --strip-args 'printf %s\n%s arg1 arg2' ) \
     <( echo -e "\n\nprintf\n\n\narg1\narg2" ) \
     "passing command and arguments"
 }
 
+TEST_generate_sha() {
```

Comment:
No sure if I understand the test:
- it seems to test `common::sha`
- we have the 2 test files as fixtures in `$TESTDATA` but still generate them on the fly
- kinda reimplement parts of `common::sha` [in the test](https://github.com/kubernetes/release/pull/849/files#diff-5b149885a8a7f68f210c00d9e5bea985R73) and the [assert on that implementation](https://github.com/kubernetes/release/pull/849/files#diff-5b149885a8a7f68f210c00d9e5bea985R89-R92)
- we don't have any test in place which actually tests if the shasum files would be uploaded when they exist, which I believe is the interesting bit

I don't want to block on all that, esp. because I believe testing `update_github_release` will be a bit painful. But I guess it will force us to refactor that func into something that is easier to test and easier to reason about.

In general, we (esp. myself) could do a better job on documenting the test cases by at least stating what we are intending to test and why.

Reply:
Great catch, @hoegaarden! Some of that test was detritus from copy/pasting the other one.

> * it seems to test `common::sha`

Fixed up in recent commit to actually test `common::sha`

> * we have the 2 test files as fixtures in `$TESTDATA` but still generate them on the fly

We're actually generating stuff in the `tmp_dir` and checking it against `TESTDATA/common/shas`.

> * kinda reimplement parts of `common::sha` [in the test](https://github.com/kubernetes/release/pull/849/files#diff-5b149885a8a7f68f210c00d9e5bea985R73) and the [assert on that implementation](https://github.com/kubernetes/release/pull/849/files#diff-5b149885a8a7f68f210c00d9e5bea985R89-R92)

Fixed in recent commit.

> * we don't have any test in place which actually tests if the shasum files would be uploaded when they exist, which I believe is the interesting bit
I don't want to block on all that, esp. because I believe testing update_github_release will be a bit painful. But I guess it will force us to refactor that func into something that is easier to test and easier to reason about.

Agreed. Let's not block on that. I explicitly didn't try to take it on here, because it felt hairy and deserving of its' own PR.
I've opened https://github.com/kubernetes/release/issues/855 for someone to work on that separately.